File: LeetCode/210. 课程表 II.cpp
class Solution {
private:
    vector<int> indegree;
    vector<int> result;
    vector<vector<int>> m;
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        indegree.resize(numCourses);
        m.resize(numCourses);
        for(auto& sub:prerequisites){
            indegree[sub[0]]++;
            m[sub[1]].push_back(sub[0]);
        }
        queue<int> q;
        for(int i=0;i<indegree.size();i++){
            if(indegree[i]==0) q.push(i);
        }
        while(!q.empty()){
            int cur=q.front();
            result.push_back(cur);
            q.pop();
            for(int neighbor:m[cur]){
                indegree[neighbor]--;
                if(indegree[neighbor]==0){
                    q.push(neighbor);
                }
            }
        }
        if(result.size()!=numCourses) return {};
        return result;
    }
};

File: LeetCode/210. 课程表 II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        result = []
        visited = [False for i in range(numCourses)]
        onPath = [False for i in range(numCourses)]
        hasCircle = False
        graph = [[] for i in range(numCourses)]
        for [cur, pre] in prerequisites:
            graph[pre].append(cur)
        def traverse(start: int):
            nonlocal onPath, visited, result, hasCircle, graph
            if onPath[start]:
                hasCircle = True
            if hasCircle or visited[start]:
                return
            onPath[start] = True
            visited[start] = True
            for x in graph[start]:
                traverse(x)
            result.append(start)
            onPath[start] = False
        for i in range(numCourses):
            traverse(i)
        return [] if hasCircle else result[::-1]
